### PR TITLE
feat(db): add checkpoint v2 for db inconsistency

### PR DESCRIPTION
**Pull request by guoquanwu**
**What does this PR do?**

Add checkpoint v2 in case of ungraceful shutdown.

**Why are these changes required?**

**This PR has been tested by:**
- Unit Tests
- Manual Testing

### CODE REVIEW

File: chainbase/src/main/java/org/tron/core/db/TronDatabase.java [hunk 1]
```diff
@@ -55,7 +52,7 @@ protected TronDatabase(String dbName) {
           CommonParameter.getInstance().getStorage().getDbDirectory()).toString();
       dbSource =
           new RocksDbDataSourceImpl(parentName, dbName, CommonParameter.getInstance()
-              .getRocksDBCustomSettings());
+              .getRocksDBCustomSettings(), getDirectComparator());
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db/TronDatabase.java:55]: why this?

File: framework/src/main/java/org/tron/tool/litefullnode/LiteFullNodeTool.java [hunk 2]
```diff
@@ -197,20 +200,46 @@ private void split(String sourceDir, String destDir, List<String> dbs) throws IO
...
+              if (destDbs != null && destDbs.contains(dbName)) {
+                DBInterface destDb = DbTool.getDB(destDir, dbName);
+                if (realValue != null) {
+                  destDb.put(realKey, realValue);
+                } else {
+                  destDb.delete(realKey);
+                }
+              }
+            }
+          }
```
Comment (halibobo1205) [inline framework/src/main/java/org/tron/tool/litefullnode/LiteFullNodeTool.java:224]: extract method .

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java [hunk 3]
```diff
@@ -414,6 +558,56 @@ public void check() {
...
+        } else {
+          dbMap.get(db).getHead().remove(realKey);
+        }
+      }
+
+      dbs.forEach(db -> db.getHead().getRoot().merge(db.getHead()));
+      retreat();
+      checkPointV2Store.close();
+      logger.info("checkpoint recover success, block number{}", blockNumber);
+    }
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java:603]: extract method .

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java [hunk 4]
```diff
@@ -360,25 +409,120 @@ private void createCheckpoint() {
...
+          DB<Key, Value> keyValueDB = snapshot.getDb();
+          for (Map.Entry<Key, Value> e : keyValueDB) {
+            Key k = e.getKey();
+            Value v = e.getValue();
+            batch.put(WrappedByteArray.of(Bytes.concat(simpleEncode(dbName), k.getBytes())),
+                WrappedByteArray.of(v.encode()));
+            if ("block".equals(db.getDbName())) {
+              currentBlockNum = new BlockCapsule(v.getBytes()).getNum();
+            }
+          }
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java:436]: extract method .

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java [hunk 5]
```diff
@@ -360,25 +394,129 @@ private void createCheckpoint() {
...
+      return;
+    }
+    List<String> cpList = getCheckpointList();
+    if (cpList == null) {
+      return;
+    }
+    if (cpList.size() < 3) {
+      return;
+    }
+    for (String cp: cpList.subList(0,3)) {
```
Comment (guoquanwu) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java:487]: 3   --->   length - 3

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java [hunk 6]
```diff
@@ -307,8 +344,22 @@ public void flush() {
...
+          Metrics.histogramObserve(requestTimer);
+
+          Histogram.Timer createTimer = Metrics.histogramStartTimer(
+              MetricKeys.Histogram.DB_FLUSH, "create");
+          createCheckpoint();
+          Metrics.histogramObserve(createTimer);
+        } else {
+          Histogram.Timer createV2 = Metrics.histogramStartTimer(
+              MetricKeys.Histogram.DB_FLUSH, "create2");
+          createCheckpointV2();
```
Comment (guoquanwu) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java:360]: `createv2` merge to `create`

**Review by halibobo1205 — APPROVED** [review]

**Pull request merged**